# Sensor Fault Detection

**Brief:** In electronics, a **wafer** (also called a slice or substrate) is a thin slice of semiconductor, such as a crystalline silicon (c-Si), used for the fabrication of integrated circuits and, in photovoltaics, to manufacture solar cells. The wafer serves as the substrate(serves as foundation for contruction of other components) for microelectronic devices built in and upon the wafer. 

It undergoes many microfabrication processes, such as doping, ion implantation, etching, thin-film deposition of various materials, and photolithographic patterning. Finally, the individual microcircuits are separated by wafer dicing and packaged as an integrated circuit.

## Problem Statement

**Data:** Wafers data


**Problem Statement:** Wafers are predominantly used to manufacture solar cells and are located at remote locations in bulk and they themselves consist of few hundreds of sensors. Wafers are fundamental of photovoltaic power generation, and production thereof requires high technology. Photovoltaic power generation system converts sunlight energy directly to electrical energy.

The motto behind figuring out the faulty wafers is to obliterate the need of having manual man-power doing the same. And make no mistake when we're saying this, even when they suspect a certain wafer to be faulty, they had to open the wafer from the scratch and deal with the issue, and by doing so all the wafers in the vicinity had to be stopped disrupting the whole process and stuff anf this is when that certain wafer was indeed faulty, however, when their suspicion came outta be false negative, then we can only imagine the waste of time, man-power and ofcourse, cost incurred.

**Solution:** Data fetched by wafers is to be passed through the machine learning pipeline and it is to be determined whether the wafer at hand is faulty or not apparently obliterating the need and thus cost of hiring manual labour.

### Import Required Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [10]:
file_path = r'wafer_23012020_041211.csv'
wafer_df = pd.read_csv(file_path)
print(wafer_df.shape)
wafer_df.head()

(100, 592)


Unnamed: 0  Sensor-1  Sensor-2   Sensor-3   Sensor-4  Sensor-5  Sensor-6  \
0  Wafer-801   2968.33   2476.58  2216.7333  1748.0885    1.1127     100.0   
1  Wafer-802   2961.04   2506.43  2170.0666  1364.5157    1.5447     100.0   
2  Wafer-803   3072.03   2500.68  2205.7445  1363.1048    1.0518     100.0   
3  Wafer-804   3021.83   2419.83  2205.7445  1363.1048    1.0518     100.0   
4  Wafer-805   3006.95   2435.34  2189.8111  1084.6502    1.1993     100.0   

   Sensor-7  Sensor-8  Sensor-9  ...  Sensor-582  Sensor-583  Sensor-584  \
0   97.5822    0.1242    1.5300  ...         NaN      0.5004      0.0120   
1   96.7700    0.1230    1.3953  ...         NaN      0.4994      0.0115   
2  101.8644    0.1220    1.3896  ...         NaN      0.4987      0.0118   
3  101.8644    0.1220    1.4108  ...         NaN      0.4934      0.0123   
4  104.8856    0.1234    1.5094  ...         NaN      0.4987      0.0145   

   Sensor-585  Sensor-586  Sensor-587  Sensor-588  Sensor-589  Sensor-590  \
0      0.0033      2.4069      0.0545      0.0184      0.0055     33.7876   
1      0.0031      2.3020      0.0545      0.0184      0.0055     33.7876   
2      0.0036      2.3719      0.0545      0.0184      0.0055     33.7876   
3      0.0040      2.4923      0.0545      0.0184      0.0055     33.7876   
4      0.0041      2.8991      0.0545      0.0184      0.0055     33.7876   

   Good/Bad  
0        -1  
1         1  
2        -1  
3        -1  
4        -1  

[5 rows x 592 columns]

In [11]:
wafer_df.describe()

Sensor-1     Sensor-2     Sensor-3     Sensor-4   Sensor-5  \
count    99.000000   100.000000    97.000000    97.000000  97.000000   
mean   3017.301212  2487.180300  2202.168281  1484.362181   1.180367   
std      71.819707    66.954212    30.350606   460.985871   0.349654   
min    2825.670000  2254.990000  2114.666700   978.783200   0.753100   
25%    2973.040000  2446.595000  2189.966700  1111.543600   0.837300   
50%    3004.390000  2493.890000  2200.988900  1244.289900   1.156900   
75%    3070.385000  2527.525000  2213.211100  1963.801600   1.383000   
max    3221.210000  2664.520000  2315.266700  2363.641200   2.207300   

       Sensor-6    Sensor-7   Sensor-8    Sensor-9   Sensor-10  ...  \
count      97.0   97.000000  97.000000  100.000000  100.000000  ...   
mean      100.0   97.449088   0.122195    1.461516    0.000243  ...   
std         0.0    5.553324   0.002006    0.071300    0.010610  ...   
min       100.0   83.423300   0.116000    1.317900   -0.027900  ...   
25%       100.0   95.108900   0.120800    1.407375   -0.006925  ...   
50%       100.0   99.513300   0.122200    1.453700    0.001000  ...   
75%       100.0  101.457800   0.123400    1.507425    0.008125  ...   
max       100.0  107.152200   0.126200    1.641100    0.025000  ...   

       Sensor-582  Sensor-583  Sensor-584  Sensor-585  Sensor-586  Sensor-587  \
count   34.000000  100.000000  100.000000  100.000000  100.000000  100.000000   
mean    74.331709    0.499390    0.013615    0.003549    2.727297    0.023510   
std     41.857728    0.003431    0.004344    0.000873    0.875848    0.011991   
min     20.309100    0.492500    0.007600    0.002100    1.515200    0.009900   
25%     47.356000    0.497300    0.011300    0.003075    2.270425    0.013400   
50%     65.127550    0.499400    0.012750    0.003400    2.546400    0.021800   
75%     99.419050    0.501525    0.014700    0.003825    2.953750    0.028025   
max    223.101800    0.508700    0.043700    0.008900    8.816000    0.054500   

       Sensor-588  Sensor-589  Sensor-590    Good/Bad  
count  100.000000  100.000000  100.000000  100.000000  
mean     0.014875    0.004685   77.430241   -0.880000  
std      0.007557    0.002527   55.106166    0.477367  
min      0.004800    0.001700   20.309100   -1.000000  
25%      0.009475    0.002700   33.787600   -1.000000  
50%      0.013900    0.003850   62.059500   -1.000000  
75%      0.019200    0.005900  104.303400   -1.000000  
max      0.040100    0.015000  223.101800    1.000000  

[8 rows x 591 columns]

### Insights
Many columns seem to have 0 standard deviation and many outliers too. So need many steps of preprocessing before modelling

In [29]:
wafer_df.isna().sum().sum() 

2306

In [23]:
na_columns = list(wafer_df.columns[wafer_df.isna().sum()>0])
print(len(na_columns))

167


#### Insights
we have 167 columns with na values and 2306 cells with na values


In [33]:
wafer_df.isna().sum().sum()/(wafer_df.shape[0]*(wafer_df.shape[1]-1))

0.03901861252115059

### Insights
Close to 4 percent of cells are missing so imputation would need to be done properly

In [25]:
wafer_df['Good/Bad'].value_counts()

Good/Bad
-1    94
 1     6
Name: count, dtype: int64

### Insights
Tarrget column is heavily imbalanced so resampling would also be required

In [27]:
# Checking Missing columns in target column
wafer_df['Good/Bad'].isna().sum()

0

### Insights
we have zero rows with target columns missing.

## Visualization

In [66]:
import random
def plot_random_50_columns(dataframe):
    """
    Plots distplot for 50 columns which are selected randomly
    """
    # Select 50 random sensors

    random_50_sensors_idx = set()
    total_columns = len(wafer_df.columns)
    num_columns = 50
    
    while len(random_50_sensors_idx)<num_columns:
        random_index = random.randint(0, total_columns - 1)
        random_50_sensors_idx.add(random_index)
    print(random_50_sensors_idx)
    random_50_sensors_idx
    

In [67]:
plot_random_50_columns(wafer_df)

592
[393, 412, 221, 441, 300, 340, 414, 65, 145, 537, 356, 193, 19, 571, 18, 557, 194, 178, 188, 265, 502, 462, 62, 71, 191, 307, 44, 37, 279, 361, 212, 237, 456, 109, 234, 357, 149, 59, 257, 417, 534, 271, 172, 440, 297, 266, 424, 55, 265, 197, 155]
